src/memray/_memray/record_writer.h

```diff
@@ -1,71 +1,69 @@
 #pragma once
 
-#include <cerrno>
-#include <climits>
-#include <cstring>
-#include <memory>
-#include <mutex>
+#include <limits>
 #include <string>
 #include <type_traits>
 #include <unistd.h>
 
-#include "records.h"
 #include "sink.h"
 
 namespace memray::tracking_api {
+
 class RecordWriter
 {
   public:
-    explicit RecordWriter(
-            std::unique_ptr<memray::io::Sink> sink,
-            const std::string& command_line,
-            bool native_traces);
-    void setMainTidAndSkippedFrames(thread_id_t main_tid, size_t skipped_frames_on_main_tid);
+    virtual ~RecordWriter() = default;
 
     RecordWriter(RecordWriter& other) = delete;
     RecordWriter(RecordWriter&& other) = delete;
     void operator=(const RecordWriter&) = delete;
     void operator=(RecordWriter&&) = delete;
 
+    virtual bool writeRecord(const MemoryRecord& record) = 0;
```

Comment:
Discussed offline: this doesn't really work nicely because of this:
https://github.com/bloomberg/memray/blob/1e5ef53c2f6d068e14e8c19aedb22d92b77a0b4c/src/memray/_memray/tracking_api.cpp#L426

We treat `Tracker` as a singleton, and we store a pointer to the created singleton instance to be referenced later. For that to work, we need to know the concrete type of the tracker, which means that templatizing it based on the type of the writer doesn't really work nicely.

I suspect that the overhead of the `virtual` calls is fairly low, and architecturally they don't seem very easy to avoid, so I'd rather keep it the way I've written it unless profiling shows it to be a significant bottleneck.